File: backend/app/schemas/nr1.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.nr1 import (
    NR1ActionStatus,
    NR1RiskCategory,
    NR1RiskSource,
    NR1RiskStatus,
    NR1TrainingStatus,
    NR1WorkspaceStatus,
)


class NR1WorkspaceUpdate(BaseModel):
    organization_name: str | None = Field(default=None, min_length=2, max_length=160)
    unit_name: str | None = Field(default=None, min_length=2, max_length=160)
    gro_owner_name: str | None = Field(default=None, min_length=2, max_length=160)
    scope_statement: str | None = Field(default=None, max_length=4000)
    notes: str | None = Field(default=None, max_length=4000)
    status: NR1WorkspaceStatus | None = None


class NR1RiskItemCreate(BaseModel):
    title: str = Field(min_length=3, max_length=160)
    description: str | None = Field(default=None, max_length=4000)
    category: NR1RiskCategory
    source: NR1RiskSource = NR1RiskSource.MANUAL
    status: NR1RiskStatus = NR1RiskStatus.OPEN
    severity: int = Field(ge=1, le=5)
    likelihood: int = Field(ge=1, le=5)
    owner_label: str | None = Field(default=None, max_length=160)
    due_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    is_template: bool = False


class NR1RiskItemUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=160)
    description: str | None = Field(default=None, max_length=4000)
    category: NR1RiskCategory | None = None
    source: NR1RiskSource | None = None
    status: NR1RiskStatus | None = None
    severity: int | None = Field(default=None, ge=1, le=5)
    likelihood: int | None = Field(default=None, ge=1, le=5)
    owner_label: str | None = Field(default=None, max_length=160)
    due_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    is_template: bool | None = None
    reviewed_at: datetime | None = None


class NR1ActionItemCreate(BaseModel):
    title: str = Field(min_length=3, max_length=160)
    description: str | None = Field(default=None, max_length=4000)
    owner_label: str | None = Field(default=None, max_length=160)
    status: NR1ActionStatus = NR1ActionStatus.PLANNED
    progress_percent: int = Field(default=0, ge=0, le=100)
    due_on: date | None = None
    completed_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    risk_item_id: str | None = None
    is_template: bool = False


class NR1ActionItemUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=160)
    description: str | None = Field(default=None, max_length=4000)
    owner_label: str | None = Field(default=None, max_length=160)
    status: NR1ActionStatus | None = None
    progress_percent: int | None = Field(default=None, ge=0, le=100)
    due_on: date | None = None
    completed_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    risk_item_id: str | None = None
    is_template: bool | None = None


class NR1TrainingItemCreate(BaseModel):
    title: str = Field(min_length=3, max_length=160)
    audience: str | None = Field(default=None, max_length=160)
    status: NR1TrainingStatus = NR1TrainingStatus.PENDING
    frequency_days: int | None = Field(default=None, ge=1, le=3650)
    required_hours: float | None = Field(default=None, ge=0, le=200)
    due_on: date | None = None
    completed_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    is_template: bool = False


class NR1TrainingItemUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=160)
    audience: str | None = Field(default=None, max_length=160)
    status: NR1TrainingStatus | None = None
    frequency_days: int | None = Field(default=None, ge=1, le=3650)
    required_hours: float | None = Field(default=None, ge=0, le=200)
    due_on: date | None = None
    completed_on: date | None = None
    notes: str | None = Field(default=None, max_length=4000)
    is_template: bool | None = None


class NR1WorkspaceResponse(BaseModel):
    id: str
    company_user_id: str
    organization_name: str
    unit_name: str | None
    gro_owner_name: str | None
    scope_statement: str | None
    notes: str | None
    status: NR1WorkspaceStatus
    last_reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class NR1RiskItemResponse(BaseModel):
    id: str
    workspace_id: str
    title: str
    description: str | None
    category: NR1RiskCategory
    source: NR1RiskSource
    status: NR1RiskStatus
    severity: int
    likelihood: int
    risk_score: int
    owner_label: str | None
    due_on: date | None
    notes: str | None
    is_template: bool
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class NR1ActionItemResponse(BaseModel):
    id: str
    workspace_id: str
    risk_item_id: str | None
    title: str
    description: str | None
    owner_label: str | None
    status: NR1ActionStatus
    progress_percent: int
    due_on: date | None
    completed_on: date | None
    notes: str | None
    is_template: bool
    is_overdue: bool
    created_at: datetime
    updated_at: datetime


class NR1TrainingItemResponse(BaseModel):
    id: str
    workspace_id: str
    title: str
    audience: str | None
    status: NR1TrainingStatus
    frequency_days: int | None
    required_hours: float | None
    due_on: date | None
    completed_on: date | None
    notes: str | None
    is_template: bool
    is_overdue: bool
    created_at: datetime
    updated_at: datetime


class NR1SummaryResponse(BaseModel):
    participant_count: int
    minimum_participants: int
    suppressed: bool
    signal_count: int | None
    average_intensity: float | None
    average_anxiety: float | None
    average_stress: float | None
    high_risk_signal_count: int | None
    open_risk_count: int
    mitigating_risk_count: int
    controlled_risk_count: int
    open_action_count: int
    overdue_action_count: int
    training_count: int
    overdue_training_count: int
    template_count: int
    summary: str
    current_state: str
    privacy_note: str
    generated_at: datetime


class NR1WorkspaceOverviewResponse(BaseModel):
    workspace: NR1WorkspaceResponse
    summary: NR1SummaryResponse
    risks: list[NR1RiskItemResponse]
    actions: list[NR1ActionItemResponse]
    trainings: list[NR1TrainingItemResponse]

File: frontend/src/lib/nr1.ts
import { apiRequest } from "@/lib/api";

export type Nr1WorkspaceStatus = "DRAFT" | "IN_REVIEW" | "ACTIVE" | "MONITORING" | "PAUSED";
export type Nr1RiskCategory = "PSYCHOSOCIAL" | "ORGANIZATIONAL" | "ERGONOMIC" | "ENVIRONMENTAL" | "SAFETY";
export type Nr1RiskSource = "TEMPLATE" | "MANUAL" | "EMPLOYEE_FEEDBACK" | "EMOTIONAL_SIGNAL" | "INCIDENT";
export type Nr1RiskStatus = "OPEN" | "MITIGATING" | "CONTROLLED" | "MONITORING" | "ARCHIVED";
export type Nr1ActionStatus = "PLANNED" | "IN_PROGRESS" | "BLOCKED" | "COMPLETED" | "CANCELED";
export type Nr1TrainingStatus = "PENDING" | "SCHEDULED" | "COMPLETED" | "OVERDUE" | "CANCELED";

export type Nr1Workspace = {
  id: string;
  company_user_id: string;
  organization_name: string;
  unit_name: string | null;
  gro_owner_name: string | null;
  scope_statement: string | null;
  notes: string | null;
  status: Nr1WorkspaceStatus;
  last_reviewed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type Nr1RiskItem = {
  id: string;
  workspace_id: string;
  title: string;
  description: string | null;
  category: Nr1RiskCategory;
  source: Nr1RiskSource;
  status: Nr1RiskStatus;
  severity: number;
  likelihood: number;
  risk_score: number;
  owner_label: string | null;
  due_on: string | null;
  notes: string | null;
  is_template: boolean;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type Nr1ActionItem = {
  id: string;
  workspace_id: string;
  risk_item_id: string | null;
  title: string;
  description: string | null;
  owner_label: string | null;
  status: Nr1ActionStatus;
  progress_percent: number;
  due_on: string | null;
  completed_on: string | null;
  notes: string | null;
  is_template: boolean;
  is_overdue: boolean;
  created_at: string;
  updated_at: string;
};

export type Nr1TrainingItem = {
  id: string;
  workspace_id: string;
  title: string;
  audience: string | null;
  status: Nr1TrainingStatus;
  frequency_days: number | null;
  required_hours: number | null;
  due_on: string | null;
  completed_on: string | null;
  notes: string | null;
  is_template: boolean;
  is_overdue: boolean;
  created_at: string;
  updated_at: string;
};

export type Nr1Summary = {
  participant_count: number;
  minimum_participants: number;
  suppressed: boolean;
  signal_count: number | null;
  average_intensity: number | null;
  average_anxiety: number | null;
  average_stress: number | null;
  high_risk_signal_count: number | null;
  open_risk_count: number;
  mitigating_risk_count: number;
  controlled_risk_count: number;
  open_action_count: number;
  overdue_action_count: number;
  training_count: number;
  overdue_training_count: number;
  template_count: number;
  summary: string;
  current_state: string;
  privacy_note: string;
  generated_at: string;
};

export type Nr1WorkspaceOverview = {
  workspace: Nr1Workspace;
  summary: Nr1Summary;
  risks: Nr1RiskItem[];
  actions: Nr1ActionItem[];
  trainings: Nr1TrainingItem[];
};

export type Nr1WorkspaceUpdate = {
  organization_name?: string | null;
  unit_name?: string | null;
  gro_owner_name?: string | null;
  scope_statement?: string | null;
  notes?: string | null;
  status?: Nr1WorkspaceStatus | null;
};

export type Nr1RiskItemCreate = {
  title: string;
  description?: string | null;
  category: Nr1RiskCategory;
  source?: Nr1RiskSource;
  status?: Nr1RiskStatus;
  severity: number;
  likelihood: number;
  owner_label?: string | null;
  due_on?: string | null;
  notes?: string | null;
  is_template?: boolean;
};

export type Nr1RiskItemUpdate = Partial<Nr1RiskItemCreate> & {
  reviewed_at?: string | null;
};

export type Nr1ActionItemCreate = {
  title: string;
  description?: string | null;
  owner_label?: string | null;
  status?: Nr1ActionStatus;
  progress_percent?: number;
  due_on?: string | null;
  completed_on?: string | null;
  notes?: string | null;
  risk_item_id?: string | null;
  is_template?: boolean;
};

export type Nr1ActionItemUpdate = Partial<Nr1ActionItemCreate>;

export type Nr1TrainingItemCreate = {
  title: string;
  audience?: string | null;
  status?: Nr1TrainingStatus;
  frequency_days?: number | null;
  required_hours?: number | null;
  due_on?: string | null;
  completed_on?: string | null;
  notes?: string | null;
  is_template?: boolean;
};

export type Nr1TrainingItemUpdate = Partial<Nr1TrainingItemCreate>;

export async function getNr1Workspace(): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>("/nr1/workspace");
}

export async function updateNr1Workspace(input: Nr1WorkspaceUpdate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>("/nr1/workspace", {
    method: "PATCH",
    body: JSON.stringify(input)
  });
}

export async function createNr1RiskItem(input: Nr1RiskItemCreate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>("/nr1/risks", {
    method: "POST",
    body: JSON.stringify(input)
  });
}

export async function updateNr1RiskItem(riskId: string, input: Nr1RiskItemUpdate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/risks/${riskId}`, {
    method: "PATCH",
    body: JSON.stringify(input)
  });
}

export async function deleteNr1RiskItem(riskId: string): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/risks/${riskId}`, {
    method: "DELETE"
  });
}

export async function createNr1ActionItem(input: Nr1ActionItemCreate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>("/nr1/actions", {
    method: "POST",
    body: JSON.stringify(input)
  });
}

export async function updateNr1ActionItem(actionId: string, input: Nr1ActionItemUpdate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/actions/${actionId}`, {
    method: "PATCH",
    body: JSON.stringify(input)
  });
}

export async function deleteNr1ActionItem(actionId: string): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/actions/${actionId}`, {
    method: "DELETE"
  });
}

export async function createNr1TrainingItem(input: Nr1TrainingItemCreate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>("/nr1/trainings", {
    method: "POST",
    body: JSON.stringify(input)
  });
}

export async function updateNr1TrainingItem(trainingId: string, input: Nr1TrainingItemUpdate): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/trainings/${trainingId}`, {
    method: "PATCH",
    body: JSON.stringify(input)
  });
}

export async function deleteNr1TrainingItem(trainingId: string): Promise<Nr1WorkspaceOverview> {
  return apiRequest<Nr1WorkspaceOverview>(`/nr1/trainings/${trainingId}`, {
    method: "DELETE"
  });
}
